### PR TITLE
AzureMonitor: move metric deep link code to backend

### LINKED ISSUE #32322: AzureMonitor: Refactor data links and move to backend
In the Azure Monitor data source plugin, for Metrics and Logs queries, data links are added to the dataframes on the _frontend_ (see [processResponse](https://github.com/grafana/grafana/blob/d71c5f4c4d1f634097c3e55289f0ef5f4ae07c2b/public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts/#L102-L131)), which is suboptimal. We should move these into the data source backend, like  [Azure Resource Graph](https://github.com/grafana/grafana/blob/d71c5f4c4d1f634097c3e55289f0ef5f4ae07c2b/pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go/#L199-L212).

This should happen for both Metrics and Logs queries, but as mentioned in https://github.com/grafana/grafana/issues/36242#issuecomment-914273449, we may need to change the URL format for Logs.

 - [x] Metrics
 - [ ] Logs - link construction is async to fetch additional details, which may change once resource-centric queries are introduced

### PR DESCRIPTION
**Pull request by iwysiu**
Related to #32322

Notes: I generated the new links with Sprintf, I can generate them with json.Marshal if it's prefered, but the nested documents for that will end up being somewhat longwinded.

### CODE REVIEW

File: pkg/tsdb/azuremonitor/azuremonitor-datasource.go [hunk 1]
```diff
@@ -299,10 +297,46 @@ func (e *AzureMonitorDatasource) parseResponse(amr AzureMonitorResponse, query *
...
+	timespan = url.QueryEscape(timespan)
+
+	id := fmt.Sprintf("/subscriptions/%v/resourceGroups/%v/providers/%v/%v",
+		query.UrlComponents["subscription"],
+		query.UrlComponents["resourceGroup"],
+		query.UrlComponents["metricDefinition"],
+		query.UrlComponents["resourceName"],
+	)
+	chartDef := fmt.Sprintf(
+		"{\"v2charts\":[{\"metrics\":[{\"resourceMetadata\":{\"id\":%q},\"name\":%q,\"aggregationType\":%v,\"namespace\":%q,\"metricVisualization\":{\"displayName\":%q,\"resourceDisplayName\":%q}}]}]}",
```
Comment (andresmgot) [inline pkg/tsdb/azuremonitor/azuremonitor-datasource.go:329]: this is a bit error prone, I would define the struct so you can `json.Marshall` it. (the same for `timespan` while you are on it)

File: pkg/tsdb/azuremonitor/azuremonitor-datasource_test.go [hunk 2]
```diff
@@ -547,3 +587,50 @@ func TestAzureMonitorCreateRequest(t *testing.T) {
...
+	queries, err := datasource.buildQueries(tsdbQuery, dsInfo)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(queries), "Expected 1 query to be built, got: %v", len(queries))
+
+	expected := `http://ds/#blade/Microsoft_Azure_MonitoringMetrics/Metrics.ReactView/Referer/MetricsExplorer/` +
+		`TimeContext/%7B%22absolute%22%3A%7B%22startTime%22%3A%222018-03-15T13%3A00%3A00Z%22%2C%22endTime%22%3A%222018-03-15T13%3A34%3A00Z%22%7D%7D/` +
+		`ChartDefinition/%7B%22v2charts%22%3A%5B%7B%22metrics%22%3A%5B%7B%22resourceMetadata%22%3A%7B%22id%22%3A%22%2Fsubscriptions%2F12345678-aaaa-bbbb-cccc-123456789abc%2FresourceGroups%2Fgrafanastaging%2Fproviders%2FMicrosoft.Compute%2FvirtualMachines%2Fgrafana%22%7D%2C` +
+		`%22name%22%3A%22Percentage+CPU%22%2C%22aggregationType%22%3A4%2C%22namespace%22%3A%22Microsoft.Compute-virtualMachines%22%2C%22metricVisualization%22%3A%7B%22displayName%22%3A%22Percentage+CPU%22%2C%22resourceDisplayName%22%3A%22grafana%22%7D%7D%5D%7D%5D%7D`
+	actual := getQueryUrl(queries[0], "http://ds")
+	require.Equal(t, expected, actual)
```
Comment (andresmgot) [inline pkg/tsdb/azuremonitor/azuremonitor-datasource_test.go:635]: isn't this test already covered with the ones above?

**Review by andresmgot — COMMENTED** [review]
Nicely done! I will let @joshhunt test this IRL. I just have some minor comments.

BTW, this doesn't fully fix #32322, this is the "Metrics" part. You will need to do something similar for logs (`azure-log-analytics` in the code) in another PR. Related to that, check [this issue](https://github.com/grafana/grafana/issues/36242#issuecomment-914273449) since it seems that it's not clear how to encode the URL for logs.

File: pkg/tsdb/azuremonitor/azuremonitor-datasource.go [hunk 3]
```diff
@@ -299,10 +297,46 @@ func (e *AzureMonitorDatasource) parseResponse(amr AzureMonitorResponse, query *
 			frame.SetRow(i, point.TimeStamp, value)
 		}
 
-		frames = append(frames, frame)
+		frameWithLink := addConfigData(*frame, getQueryUrl(query, azurePortalUrl))
+		frames = append(frames, &frameWithLink)
+	}
+
+	return frames
```
[the remark this answers is not in the document]
Comment (joshhunt) [reply pkg/tsdb/azuremonitor/azuremonitor-datasource.go:304]: I had this commented drafted up for executeQuery:

> Not sure if I'm reading this correctly, but it looks like executeQuery will no longer return any errors, and queries now will just silently fail if there's a problem with them?

Is this what you're referring to?

[the remark this answers is not in the document]
Comment (andresmgot) [reply pkg/tsdb/azuremonitor/azuremonitor-datasource.go:304]: Yes but no. `executeQuery` adds the error to the frame (see for example the line 183) so the `err` returned in the function was always `nil`.

**Review by iwysiu — COMMENTED** [review]
Also edited the pull request comment

File: pkg/tsdb/azuremonitor/azuremonitor-datasource.go [hunk 4]
```diff
@@ -299,12 +302,84 @@ func (e *AzureMonitorDatasource) parseResponse(amr AzureMonitorResponse, query *
...
+		"v2charts": []interface{}{
+			map[string]interface{}{
+				"metrics": []interface{}{
+					struct {
+						ResourceMetadata    map[string]string `json:"resourceMetadata"`
+						Name                string            `json:"name"`
+						AggregationType     int               `json:"aggregationType"`
+						Namespace           string            `json:"namespace"`
+						MetricVisualization interface{}       `json:"metricVisualization"`
+					}{
```
Comment (andresmgot) [inline pkg/tsdb/azuremonitor/azuremonitor-datasource.go:356]: can you move this definition to `types.go`?

**Review by andresmgot — APPROVED** [review]
LGTM! just one minor comment left. (I haven't tried this IRL)

**Review by andresmgot — APPROVED** [review]
:+1:

**Comment by andresmgot** [comment]
You can merge this once you are ready @iwysiu (unless @joshhunt says otherwise). I have added the backport label and milestone. For more info about merging PRs see [this doc](https://github.com/grafana/grafana/blob/main/contribute/merge-pull-request.md).

**Pull request merged**